cmd/kubeadm/app/cmd/upgrade/diff.go

```diff
@@ -0,0 +1,139 @@
+	newK8sVersionStr              string
+	contextLines                  int
+	parent                        *cmdUpgradeFlags
+}
+
+var (
+	defaultAPIServerManifestPath         = constants.GetStaticPodFilepath(constants.KubeAPIServer, constants.GetStaticPodDirectory())
+	defaultControllerManagerManifestPath = constants.GetStaticPodFilepath(constants.KubeControllerManager, constants.GetStaticPodDirectory())
+	defaultSchedulerManifestPath         = constants.GetStaticPodFilepath(constants.KubeScheduler, constants.GetStaticPodDirectory())
+)
+
+// NewCmdDiff returns the cobra command for `kubeadm upgrade diff`
+func NewCmdDiff(parentflags *cmdUpgradeFlags) *cobra.Command {
+	flags := &diffFlags{
+		parent: parentflags,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "diff [version]",
+		Short: "Show what differences would be applied to existing static pod manifests. See also: kubeadm upgrade apply --dry-run",
+		Run: func(cmd *cobra.Command, args []string) {
+			kubeadmutil.CheckErr(runDiff(flags, args))
+		},
+	}
+
+	cmd.Flags().StringVar(&flags.apiServerManifestPath, "api-server-manifest", defaultAPIServerManifestPath, "path to API server manifest")
+	cmd.Flags().StringVar(&flags.controllerManagerManifestPath, "controller-manager-manifest", defaultControllerManagerManifestPath, "path to controller manifest")
+	cmd.Flags().StringVar(&flags.schedulerManifestPath, "scheduler-manifest", defaultSchedulerManifestPath, "path to scheduler manifest")
+	cmd.Flags().IntVarP(&flags.contextLines, "context-lines", "c", 3, "How many lines of context in the diff")
+
+	return cmd
+}
+
+func runDiff(flags *diffFlags, args []string) error {
+
+	// If the version is specified in config file, pick up that value.
+	glog.V(1).Infof("fetching configuration from file", flags.parent.cfgPath)
+	cfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(flags.parent.cfgPath, &kubeadmv1alpha2.MasterConfiguration{})
+	if err != nil {
+		return err
+	}
+
+	if cfg.KubernetesVersion != "" {
+		flags.newK8sVersionStr = cfg.KubernetesVersion
+	}
+
+	// If the new version is already specified in config file, version arg is optional.
+	if flags.newK8sVersionStr == "" {
+		if err := cmdutil.ValidateExactArgNumber(args, []string{"version"}); err != nil {
+			return err
+		}
+	}
+
+	// If option was specified in both args and config file, args will overwrite the config file.
+	if len(args) == 1 {
+		flags.newK8sVersionStr = args[0]
+	}
+
+	k8sVer, err := version.ParseSemantic(flags.newK8sVersionStr)
+	if err != nil {
+		return err
+	}
+
+	specs := controlplane.GetStaticPodSpecs(cfg, k8sVer)
+	for spec, pod := range specs {
+		var path string
+		switch spec {
+		case constants.KubeAPIServer:
+			path = flags.apiServerManifestPath
+		case constants.KubeControllerManager:
+			path = flags.controllerManagerManifestPath
+		case constants.KubeScheduler:
+			path = flags.schedulerManifestPath
+		default:
+			glog.Errorf("[diff] unknown spec %v", spec)
+			continue
```

Comment:
is it OK to skip unknown specs here?
should we `return fmt.Errorf()` instead.

Reply:
I'd prefer to err on the side of not breaking this for users. This shouldn't come up, but in case it does I don't want it to break `upgrade diff` altogether.

Reply:
sounds good.